File: android/app/src/main/java/io/talise/app/feature/ramps/IdentityVerificationView.kt
package io.talise.app.feature.ramps

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.OpenInNew
import androidx.compose.material.icons.filled.Badge
import androidx.compose.material.icons.filled.Description
import androidx.compose.material.icons.filled.PersonSearch
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material.icons.filled.Verified
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalUriHandler
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import io.talise.app.ui.components.rampCard
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Identity verification (Bridge KYC), ported 1:1 from iOS
 * `IdentityVerificationView.swift` (Profile). Lives in the ramps package so the
 * cash-out gate is self-contained on Android.
 *
 * Flow:
 *   1. Load current status (GET /api/kyc/bridge/status).
 *   2. "Verify identity" -> POST /api/kyc/bridge/start, then open the hosted
 *      KYC + Terms-of-Service links in the browser.
 *   3. Poll status while it's in review; flip to a success state on approval.
 *
 * Verifying here unlocks USD/EUR cash-out, one Bridge customer covers both
 * directions. No PII flows through Talise; Bridge runs the whole flow.
 */
class IdentityVerificationViewModel : ViewModel() {

    data class State(
        val loading: Boolean = true,
        val working: Boolean = false,          // start() in flight
        val status: KYCStatus = KYCStatus.Unverified,
        val kycUrl: String? = null,
        val tosUrl: String? = null,
        val error: String? = null,
        val polling: Boolean = false,
        /** One-shot: a URL the screen should open, cleared after consumption. */
        val openUrl: String? = null,
    )

    private val _state = MutableStateFlow(State())
    val state: StateFlow<State> = _state.asStateFlow()

    init {
        load()
    }

    fun consumeOpenUrl() {
        _state.value = _state.value.copy(openUrl = null)
    }

    private fun load() {
        viewModelScope.launch {
            try {
                val s = RampsClient.api.kycStatus()
                val status = KYCStatus.from(s.status)
                _state.value = _state.value.copy(status = status, loading = false)
                if (status.isInFlight) startPolling()
            } catch (t: Throwable) {
                val msg = RampsClient.errorText(t)
                if (msg.contains("503")) {
                    _state.value = _state.value.copy(
                        error = "Cash-out verification isn't switched on yet.",
                        loading = false,
                    )
                } else {
                    // Soft-fail to unverified; the action card lets them start.
                    _state.value = _state.value.copy(status = KYCStatus.Unverified, loading = false)
                }
            }
        }
    }

    /** Kick off (or re-fetch) the hosted KYC + ToS links and open them. */
    fun beginVerification() {
        if (_state.value.working) return
        _state.value = _state.value.copy(working = true, error = null)
        viewModelScope.launch {
            try {
                val r = RampsClient.api.kycStart(RampsEmptyBody())
                // Open the identity flow straight away; the ToS button stays
                // available below for the second step.
                _state.value = _state.value.copy(
                    status = KYCStatus.from(r.status),
                    kycUrl = r.kycUrl,
                    tosUrl = r.tosUrl,
                    openUrl = r.kycUrl ?: r.tosUrl,
                    working = false,
                )
                startPolling()
            } catch (t: Throwable) {
                val msg = RampsClient.errorText(t)
                val error = when {
                    msg.contains("503") -> "Cash-out verification isn't switched on yet. Please try again soon."
                    msg.contains("400") -> "Add an email to your account first, then verify your identity."
                    msg.contains("429") -> "Too many attempts, wait a moment and try again."
                    msg.contains("401") -> "Your session expired. Sign out and back in, then try again."
                    else -> "Couldn't start verification. Please try again."
                }
                _state.value = _state.value.copy(error = error, working = false)
            }
        }
    }

    fun refresh() {
        viewModelScope.launch {
            runCatching { RampsClient.api.kycStatus() }.onSuccess { s ->
                _state.value = _state.value.copy(status = KYCStatus.from(s.status))
            }
            // Keep current status on failure.
        }
    }

    /** Poll every 8s while verification is in review; stops on a terminal state
     *  or when the ViewModel is cleared. */
    private fun startPolling() {
        if (_state.value.polling) return
        _state.value = _state.value.copy(polling = true)
        viewModelScope.launch {
            try {
                repeat(60) {            // ~8 minutes max
                    delay(8_000)
                    val s = runCatching { RampsClient.api.kycStatus() }.getOrNull() ?: return@repeat
                    val status = KYCStatus.from(s.status)
                    _state.value = _state.value.copy(status = status)
                    if (status == KYCStatus.Approved || status == KYCStatus.Rejected) return@launch
                }
            } finally {
                _state.value = _state.value.copy(polling = false)
            }
        }
    }
}

@Composable
fun IdentityVerificationView(
    vm: IdentityVerificationViewModel = viewModel { IdentityVerificationViewModel() },
) {
    val state by vm.state.collectAsStateWithLifecycle()
    val uriHandler = LocalUriHandler.current

    // One-shot open of the hosted flow when start() hands links back.
    LaunchedEffect(state.openUrl) {
        state.openUrl?.let { url ->
            uriHandler.openUri(url)
            vm.consumeOpenUrl()
        }
    }

    Column(
        Modifier
            .fillMaxSize()
            .background(TaliseColors.bg)
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 20.dp)
            .padding(top = 8.dp, bottom = 28.dp),
        verticalArrangement = Arrangement.spacedBy(22.dp),
    ) {
        // -- Header --
        Column(Modifier.padding(top = 4.dp), verticalArrangement = Arrangement.spacedBy(6.dp)) {
            Text(
                "Identity verification",
                style = TaliseType.heading(22.sp, FontWeight.Medium),
                letterSpacing = (-0.4).sp,
                color = TaliseColors.fg,
            )
            Text(
                "A one-time check that unlocks cashing out to your bank. Your details go straight to our payments partner, Talise never stores them.",
                style = TaliseType.body(13.5.sp, FontWeight.Light),
                color = TaliseColors.fgMuted,
            )
        }

        if (state.loading) {
            LoadingCard("Checking your status…")
        } else {
            StatusCard(status = state.status, polling = state.polling)
            when (state.status) {
                KYCStatus.Approved -> ApprovedCard()
                KYCStatus.Pending -> PendingCard(
                    kycUrl = state.kycUrl,
                    tosUrl = state.tosUrl,
                    onOpen = { uriHandler.openUri(it) },
                    onRefresh = { vm.refresh() },
                )
                KYCStatus.Rejected, KYCStatus.Expired -> ActionCard(
                    retry = true,
                    working = state.working,
                    kycUrl = state.kycUrl,
                    tosUrl = state.tosUrl,
                    onOpen = { uriHandler.openUri(it) },
                    onBegin = { vm.beginVerification() },
                )
                KYCStatus.Unverified -> ActionCard(
                    retry = false,
                    working = state.working,
                    kycUrl = state.kycUrl,
                    tosUrl = state.tosUrl,
                    onOpen = { uriHandler.openUri(it) },
                    onBegin = { vm.beginVerification() },
                )
            }
            state.error?.let {
                Text(
                    it,
                    style = TaliseType.body(13.sp, FontWeight.Light),
                    color = TaliseColors.sentRedSoft,
                )
            }
        }
    }
}

// MARK: - Sections

/** A status chip row, always shown once loaded. */
@Composable
private fun StatusCard(status: KYCStatus, polling: Boolean) {
    val (icon, tint) = statusIconTint(status)
    Row(
        Modifier.fillMaxWidth().rampCard().padding(18.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(20.dp))
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(
                "STATUS",
                style = TaliseType.mono(10.sp),
                letterSpacing = 1.sp,
                color = TaliseColors.fgDim,
            )
            Text(
                status.label,
                style = TaliseType.heading(17.sp, FontWeight.Medium),
                color = TaliseColors.fg,
            )
        }
        if (polling) {
            CircularProgressIndicator(
                color = TaliseColors.greenMint,
                strokeWidth = 2.dp,
                modifier = Modifier.size(18.dp),
            )
        }
    }
}

private fun statusIconTint(status: KYCStatus): Pair<ImageVector, Color> = when (status) {
    KYCStatus.Approved -> Icons.Filled.Verified to TaliseColors.greenMint
    KYCStatus.Pending -> Icons.Filled.Schedule to TaliseColors.fgMuted
    KYCStatus.Rejected, KYCStatus.Expired -> Icons.Filled.Warning to TaliseColors.sentRedSoft
    KYCStatus.Unverified -> Icons.Filled.PersonSearch to TaliseColors.fgMuted
}

@Composable
private fun ApprovedCard() {
    Column(
        Modifier.fillMaxWidth().rampCard().padding(18.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Icon(Icons.Filled.Verified, contentDescription = null, tint = TaliseColors.greenMint, modifier = Modifier.size(16.dp))
            Text(
                "You're verified",
                style = TaliseType.heading(16.sp, FontWeight.SemiBold),
                color = TaliseColors.greenMint,
            )
        }
        Text(
            "Cash-out to your bank is unlocked. You can withdraw from any supported corridor.",
            style = TaliseType.body(13.sp, FontWeight.Light),
            color = TaliseColors.fgMuted,
        )
    }
}

@Composable
private fun PendingCard(
    kycUrl: String?,
    tosUrl: String?,
    onOpen: (String) -> Unit,
    onRefresh: () -> Unit,
) {
    Column(
        Modifier.fillMaxWidth().rampCard().padding(18.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Text(
            "We're reviewing your details. This usually takes a few minutes. You can close this screen, we'll keep checking.",
            style = TaliseType.body(13.5.sp, FontWeight.Light),
            color = TaliseColors.fgMuted,
        )
        if (kycUrl != null || tosUrl != null) {
            OpenLinksRow(kycUrl = kycUrl, tosUrl = tosUrl, onOpen = onOpen)
        }
        Box(
            Modifier
                .fillMaxWidth()
                .height(50.dp)
                .background(TaliseColors.greenMint, CircleShape)
                .clickable { onRefresh() },
            contentAlignment = Alignment.Center,
        ) {
            Text("Refresh status", style = TaliseType.body(15.sp, FontWeight.SemiBold), color = Color.Black)
        }
    }
}

/** The "start (or retry) verification" call-to-action. */
@Composable
private fun ActionCard(
    retry: Boolean,
    working: Boolean,
    kycUrl: String?,
    tosUrl: String?,
    onOpen: (String) -> Unit,
    onBegin: () -> Unit,
) {
    Column(
        Modifier.fillMaxWidth().rampCard().padding(18.dp),
        verticalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Text(
            if (retry) {
                "Your last attempt didn't go through. You can try again, make sure your name matches your government ID."
            } else {
                "You'll verify your identity and accept the terms with our payments partner. Two quick steps in your browser."
            },
            style = TaliseType.body(13.5.sp, FontWeight.Light),
            color = TaliseColors.fgMuted,
        )

        // If links already exist (start was tapped), surface them.
        if (kycUrl != null || tosUrl != null) {
            OpenLinksRow(kycUrl = kycUrl, tosUrl = tosUrl, onOpen = onOpen)
        }

        Row(
            Modifier
                .fillMaxWidth()
                .height(54.dp)
                .background(TaliseColors.greenMint.copy(alpha = if (working) 0.7f else 1f), CircleShape)
                .clickable(enabled = !working) { onBegin() },
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp, Alignment.CenterHorizontally),
        ) {
            if (working) {
                CircularProgressIndicator(color = Color.Black, strokeWidth = 2.dp, modifier = Modifier.size(16.dp))
            }
            Text(
                if (working) "Preparing…" else if (retry) "Try again" else "Verify identity",
                style = TaliseType.body(16.sp, FontWeight.SemiBold),
                color = Color.Black,
            )
        }
    }
}

/** Re-openable KYC + ToS link buttons (shown once `start` returns them). */
@Composable
private fun OpenLinksRow(
    kycUrl: String?,
    tosUrl: String?,
    onOpen: (String) -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        kycUrl?.let { LinkButton("Verify identity", Icons.Filled.Badge) { onOpen(it) } }
        tosUrl?.let { LinkButton("Review & accept terms", Icons.Filled.Description) { onOpen(it) } }
    }
}

@Composable
private fun LinkButton(title: String, icon: ImageVector, action: () -> Unit) {
    Row(
        Modifier
            .fillMaxWidth()
            .height(48.dp)
            .background(TaliseColors.surface2, RoundedCornerShape(14.dp))
            .clickable { action() }
            .padding(horizontal = 16.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Icon(icon, contentDescription = null, tint = TaliseColors.fg, modifier = Modifier.size(13.dp))
        Text(
            title,
            style = TaliseType.body(14.5.sp, FontWeight.Medium),
            color = TaliseColors.fg,
            modifier = Modifier.weight(1f),
        )
        Icon(
            Icons.AutoMirrored.Filled.OpenInNew,
            contentDescription = null,
            tint = TaliseColors.fg,
            modifier = Modifier.size(11.dp),
        )
    }
}

File: ios/Talise/Features/Profile/IdentityVerificationView.swift
import SwiftUI

/// Identity verification (Bridge KYC), surfaced from Profile.
///
/// Flow:
///   1. Load current status (GET /api/kyc/bridge/status).
///   2. "Verify identity" → POST /api/kyc/bridge/start, then open the hosted
///      KYC + Terms-of-Service links in Safari.
///   3. Poll status while it's in review; flip to a success state on approval.
///
/// Verifying here unlocks USD/EUR cash-out — one Bridge customer covers both
/// directions. No PII flows through Talise; Bridge runs the whole flow.
struct IdentityVerificationView: View {
    @Environment(\.openURL) private var openURL

    @State private var loading = true
    @State private var working = false           // start() in flight
    @State private var status: KYCStatus = .unverified
    @State private var kycUrl: String?
    @State private var tosUrl: String?
    @State private var error: String?
    @State private var polling = false
    /// The Bridge/Persona URL presented in-app via SFSafariViewController.
    @State private var safariLink: KYCLink?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 22) {
                header
                if loading {
                    loadingCard
                } else {
                    statusCard
                    switch status {
                    case .approved:
                        approvedCard
                    case .pending:
                        pendingCard
                    case .rejected, .expired:
                        actionCard(retry: true)
                    case .unverified:
                        actionCard(retry: false)
                    }
                    if let error {
                        Text(error)
                            .font(TaliseFont.body(13, weight: .light))
                            .foregroundStyle(Color(hex: 0xFF6B6B))
                            .fixedSize(horizontal: false, vertical: true)
                    }
                }
            }
            .padding(.horizontal, 20)
            .padding(.top, 8)
            .padding(.bottom, 28)
        }
        .background(TaliseColor.bg.ignoresSafeArea())
        .navigationTitle("")
        .navigationBarTitleDisplayMode(.inline)
        .task { await load() }
        // Run Bridge/Persona identity verification IN-APP (SFSafariViewController
        // = Safari's process → full WebKit + camera). Bouncing to the external
        // browser via openURL left the inquiry sections unresponsive. On dismiss
        // we re-check status so an approval flips the card immediately.
        .fullScreenCover(item: $safariLink, onDismiss: { Task { await refresh() } }) { link in
            SafariView(url: link.url).ignoresSafeArea()
        }
    }

    // MARK: - Sections

    private var header: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("Identity verification")
                .font(TaliseFont.heading(22, weight: .medium))
                .kerning(-0.4)
                .foregroundStyle(TaliseColor.fg)
            Text("A one-time check that unlocks cashing out to your bank. Your details go straight to our payments partner — Talise never stores them.")
                .font(TaliseFont.body(13.5, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
                .fixedSize(horizontal: false, vertical: true)
        }
        .padding(.top, 4)
    }

    private var loadingCard: some View {
        HStack(spacing: 12) {
            ProgressView().tint(TaliseColor.greenMint)
            Text("Checking your status…")
                .font(TaliseFont.body(14, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
            Spacer(minLength: 0)
        }
        .padding(18)
        .rampCard()
    }

    /// A status chip row, always shown once loaded.
    private var statusCard: some View {
        HStack(spacing: 12) {
            Image(systemName: statusIcon)
                .font(.system(size: 20, weight: .regular))
                .foregroundStyle(statusColor)
            VStack(alignment: .leading, spacing: 2) {
                Text("STATUS")
                    .font(TaliseFont.mono(10, weight: .regular)).kerning(1)
                    .foregroundStyle(TaliseColor.fgDim)
                Text(status.label)
                    .font(TaliseFont.heading(17, weight: .medium))
                    .foregroundStyle(TaliseColor.fg)
            }
            Spacer(minLength: 0)
            if polling { ProgressView().tint(TaliseColor.greenMint) }
        }
        .padding(18)
        .rampCard()
    }

    private var approvedCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            Label("You're verified", systemImage: "checkmark.seal.fill")
                .font(TaliseFont.heading(16, weight: .semibold))
                .foregroundStyle(TaliseColor.greenMint)
            Text("Cash-out to your bank is unlocked. You can withdraw from any supported corridor.")
                .font(TaliseFont.body(13, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(18)
        .rampCard()
    }

    private var pendingCard: some View {
        VStack(alignment: .leading, spacing: 14) {
            Text("We're reviewing your details. This usually takes a few minutes. You can close this screen — we'll keep checking.")
                .font(TaliseFont.body(13.5, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
                .fixedSize(horizontal: false, vertical: true)
            if kycUrl != nil || tosUrl != nil {
                openLinksRow
            }
            Button { Task { await refresh() } } label: {
                Text("Refresh status")
                    .font(TaliseFont.body(15, weight: .semibold)).foregroundStyle(.black)
                    .frame(maxWidth: .infinity).frame(height: 50)
                    .background(Capsule().fill(TaliseColor.greenMint))
            }
            .buttonStyle(.plain)
        }
        .padding(18)
        .rampCard()
    }

    /// The "start (or retry) verification" call-to-action.
    private func actionCard(retry: Bool) -> some View {
        VStack(alignment: .leading, spacing: 14) {
            Text(retry
                 ? "Your last attempt didn't go through. You can try again — make sure your name matches your government ID."
                 : "You'll verify your identity and accept the terms with our payments partner. Two quick steps in your browser.")
                .font(TaliseFont.body(13.5, weight: .light))
                .foregroundStyle(TaliseColor.fgMuted)
                .fixedSize(horizontal: false, vertical: true)

            // If links already exist (start was tapped), surface them.
            if kycUrl != nil || tosUrl != nil { openLinksRow }

            Button { Task { await beginVerification() } } label: {
                HStack(spacing: 8) {
                    if working { ProgressView().tint(.black) }
                    Text(working ? "Preparing…" : (retry ? "Try again" : "Verify identity"))
                }
                .font(TaliseFont.body(16, weight: .semibold)).foregroundStyle(.black)
                .frame(maxWidth: .infinity).frame(height: 54)
                .background(Capsule().fill(TaliseColor.greenMint))
            }
            .buttonStyle(.plain)
            .disabled(working)
            .opacity(working ? 0.7 : 1)
        }
        .padding(18)
        .rampCard()
    }

    /// Re-openable KYC + ToS link buttons (shown once `start` returns them).
    private var openLinksRow: some View {
        VStack(spacing: 10) {
            if let kycUrl, let url = URL(string: kycUrl) {
                linkButton("Verify identity", system: "person.text.rectangle") { safariLink = KYCLink(url: url) }
            }
            if let tosUrl, let url = URL(string: tosUrl) {
                linkButton("Review & accept terms", system: "doc.text") { safariLink = KYCLink(url: url) }
            }
        }
    }

    private func linkButton(_ title: String, system: String, _ action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: 8) {
                Image(systemName: system).font(.system(size: 13, weight: .medium))
                Text(title).font(TaliseFont.body(14.5, weight: .medium))
                Spacer(minLength: 0)
                Image(systemName: "arrow.up.right").font(.system(size: 11, weight: .semibold))
            }
            .foregroundStyle(TaliseColor.fg)
            .padding(.horizontal, 16).frame(height: 48)
            .background(RoundedRectangle(cornerRadius: 14, style: .continuous).fill(TaliseColor.surface2))
        }
        .buttonStyle(.plain)
    }

    private var statusIcon: String {
        switch status {
        case .approved: return "checkmark.seal.fill"
        case .pending:  return "clock.fill"
        case .rejected, .expired: return "exclamationmark.triangle.fill"
        case .unverified: return "person.crop.circle.badge.questionmark"
        }
    }
    private var statusColor: Color {
        switch status {
        case .approved: return TaliseColor.greenMint
        case .pending:  return TaliseColor.fgMuted
        case .rejected, .expired: return Color(hex: 0xFF6B6B)
        case .unverified: return TaliseColor.fgMuted
        }
    }

    // MARK: - Actions

    private func load() async {
        loading = true
        defer { loading = false }
        do {
            let s = try await BridgeKYCAPI.status()
            status = KYCStatus(s.status)
            if status.isInFlight { startPolling() }
        } catch APIError.status(let code, _) where code == 503 {
            error = "Cash-out verification isn't switched on yet."
        } catch {
            // Soft-fail to unverified; the action card lets them start.
            status = .unverified
        }
    }

    /// Kick off (or re-fetch) the hosted KYC + ToS links and open them.
    private func beginVerification() async {
        guard !working else { return }
        working = true
        error = nil
        defer { working = false }
        do {
            let r = try await BridgeKYCAPI.start()
            status = KYCStatus(r.status)
            kycUrl = r.kycUrl
            tosUrl = r.tosUrl
            // Open the identity flow straight away; the ToS button stays
            // available below for the second step.
            if let s = r.kycUrl, let url = URL(string: s) {
                safariLink = KYCLink(url: url)
            } else if let s = r.tosUrl, let url = URL(string: s) {
                safariLink = KYCLink(url: url)
            }
            startPolling()
        } catch APIError.status(let code, _) where code == 503 {
            self.error = "Cash-out verification isn't switched on yet. Please try again soon."
        } catch APIError.status(let code, _) where code == 400 {
            self.error = "Add an email to your account first, then verify your identity."
        } catch APIError.status(let code, _) where code == 429 {
            self.error = "Too many attempts — wait a moment and try again."
        } catch APIError.unauthorized {
            self.error = "Your session expired. Sign out and back in, then try again."
        } catch {
            self.error = "Couldn't start verification. Please try again."
        }
    }

    private func refresh() async {
        do {
            let s = try await BridgeKYCAPI.status()
            status = KYCStatus(s.status)
        } catch { /* keep current status */ }
    }

    /// Poll every 8s while verification is in review; stops on a terminal state
    /// or when the view goes away (the Task is cancelled by `.task` teardown).
    private func startPolling() {
        guard !polling else { return }
        polling = true
        Task {
            defer { polling = false }
            for _ in 0..<60 {            // ~8 minutes max
                try? await Task.sleep(nanoseconds: 8_000_000_000)
                if Task.isCancelled { return }
                do {
                    let s = try await BridgeKYCAPI.status()
                    status = KYCStatus(s.status)
                } catch { continue }
                if status == .approved || status == .rejected { return }
            }
        }
    }
}

/// Identifiable wrapper so a Bridge/Persona URL can drive `.fullScreenCover(item:)`.
private struct KYCLink: Identifiable {
    let id = UUID()
    let url: URL
}
